pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/TimePredicateFilterOptimizer.java

```diff
@@ -0,0 +1,330 @@
+        lowerInclusive = inputTimeUnit.toMillis(lowerValue) == lowerMillis;
+      }
+      Long upperValue = null;
+      boolean upperInclusive = false;
+      if (upperMillis != null) {
+        // formatToMillis(col) < millis
+        // - if (formatToMillis(millisToFormat(millis)) == millis)
+        //   -> col < millisToFormat(millis)
+        // - else (formatToMillis(millisToFormat(millis)) < millis)
+        //   -> col <= millisToFormat(millis)
+        upperValue = inputTimeUnit.convert(upperMillis, TimeUnit.MILLISECONDS);
+        upperInclusive = inputTimeUnit.toMillis(upperValue) != upperMillis;
+      }
+
+      // Step 3: Rewrite the filter function
+      String rangeString = new Range(lowerValue, lowerInclusive, upperValue, upperInclusive).getRangeString();
+      filterFunction.setOperator(FilterKind.RANGE.name());
+      filterFunction
+          .setOperands(Arrays.asList(timeConvertOperands.get(0), RequestUtils.getLiteralExpression(rangeString)));
+    } catch (Exception e) {
+      LOGGER.warn("Caught exception while optimizing TIME_CONVERT predicate: {}, skipping the optimization",
+          filterFunction, e);
+    }
+  }
+
+  /**
+   * Helper method to optimize DATE_TIME_CONVERT function with range/equality predicate to directly apply the predicate
+   * to the inner expression. Changes are applied in-place of the filter function.
+   */
+  private void optimizeDateTimeConvert(Function filterFunction, FilterKind filterKind) {
+    List<Expression> filterOperands = filterFunction.getOperands();
+    List<Expression> dateTimeConvertOperands = filterOperands.get(0).getFunctionCall().getOperands();
+    Preconditions.checkArgument(dateTimeConvertOperands.size() == 4,
+        "Exactly 4 arguments are required for DATE_TIME_CONVERT transform function");
+    Preconditions.checkArgument(
+        isStringLiteral(dateTimeConvertOperands.get(1)) && isStringLiteral(dateTimeConvertOperands.get(2))
+            && isStringLiteral(dateTimeConvertOperands.get(3)),
+        "The 2nd to 4th arguments for DATE_TIME_CONVERT transform function must be string literal");
+
+    try {
+      DateTimeFormatSpec inputFormat =
+          new DateTimeFormatSpec(dateTimeConvertOperands.get(1).getLiteral().getStringValue());
+      DateTimeFormatSpec outputFormat =
+          new DateTimeFormatSpec(dateTimeConvertOperands.get(2).getLiteral().getStringValue());
+      // SDF output format is not supported because:
+      // 1. No easy way to get the next time value (instead of simply +1 for EPOCH format)
+      // 2. Hard to calculate the bucket boundary (need to consider time zone)
+      // TODO: Support SDF output format
+      if (outputFormat.getTimeFormat() == TimeFormat.SIMPLE_DATE_FORMAT) {
+        return;
+      }
+      long granularityMillis = new DateTimeGranularitySpec(dateTimeConvertOperands.get(3).getLiteral().getStringValue())
+          .granularityToMillis();
+
+      // Step 1: Convert output range to millis range
```

Comment:
Can step 1 and step 2 code be refactored and made common between `optimizeTimeConvert `and `optimizeDateTimeConvert` functions?

Reply:
I considered using the same code for both `time_convert` and `date_time_convert`, but that would introduce unnecessary overhead for `time_convert` because it does not require SDF and bucketing support. Basically it's a trade-off between performance and code complexity